text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,559 @@
+**Python:** https://github.com/open-telemetry/opentelemetry-python/pull/278  
+**Ruby:** https://github.com/open-telemetry/opentelemetry-ruby/pull/147  
+
+# Examples
+
+It might be helpful to look at an example, written in pseudocode. Let's describe 
+a simple scenario, where `service A` responds to an HTTP request from a `client` 
+with the result of a request to `service B`.
+
+```
+client -> service A -> service B
+```
+
+Now, let's assume the `client` in the above system is version 1.0. With version 
+v2.0 of the `client`, `service A` must call `service C` instead of `service B` 
+in order to return the correct data.
+
+```
+client -> service A -> service C
+```
+
+In , we would like `service A` to decide on which backend service to call, 
+based on the client version. We would also like to trace the entire system, in 
+order to understand if requests to `service C` are slower or faster than 
+`service B`. What might `service A` look like?
+
+First, during program initialization, `service A` might set a global 
+extractor and injector which chains together baggage and tracing 
+propagation. Let's assume this tracing system is configured to use B3, 
+and has a specific propagator for that format. Initializating the propagators 
+might look like this:
+
+```php
+func InitializeOpentelemetry() {
+  // create the propagators for tracing and baggage.
+  bagExtract, bagInject = Baggage::HTTPPropagator()
+  traceExtract, traceInject = Tracer::B3Propagator()
+  
+  // chain the propagators together and make them globally available.
+  extract = Propagation::ChainHTTPExtractor(bagExtract,traceExtract)
+  inject = Propagation::ChainHTTPInjector(bagInject,traceInject)
+
+  Propagation::SetHTTPExtractor(extract)
+  Propagation::SetHTTPInjector(inject)
+}
+```
+
+These propagators can then be used in the request handler for `service A`. The 
+tracing and baggage aspects use the context object to handle state without 
+breaking the encapsulation of the functions they are embedded in.
+
+```php
+func HandleUpstreamRequest(context, request, project) -> (context) {
+  // Extract the span context. Because the extractors have been chained,
+  // both a span context and any upstream baggage have been extracted 
+  // from the request headers into the returned context.
+  extract = Propagation::GetHTTPExtractor()
+  context = extract(context, request.Headers)
+
+  // Start a span, setting the parent to the span context received from 
+  // the upstream. The new span will then be in the returned context.
+  context = Tracer::StartSpan(context, [span options])
+  
+  version = Baggage::GetBaggage( context, "client-version")
+
+  switch( version ){
+    case "v1.0":
+      data, context = FetchDataFromServiceB(context)
+    case "v2.0":
+      data, context = FetchDataFromServiceC(context)
+  }
+
+  context = request.Response(context, data)
+
+  // End the current span
+  context = Tracer::EndSpan(context)
```

Comment:
do we really have this in the spec, that End returns another context?

Reply:
it doesn't make much sense to me, certainly not for Go or other languages with explicit context, because restoring previously active span happens automatically by exiting the current lexical scope where the earlier instance of Context does not contain the current span.

It would make more sense in the language with implicit (thread local) context, but _only in combination with_ an auto-closable Scope. So this example still looks very odd to me.

Reply:
I'd be interested to see an alternative example. It seems straight forwards that ending a span could modify the contents of the context object. Assuming that the contents of an immutable object are mutable is an odd assertion. 

Aside: since work can continue after a span ends, without a change in lexical scope, I'm not sure if that last point is relevant, even for Go.

Reply:
If you want the previous context to be active, you simply keep a reference to it:
```
func (ctx1 Context) {
    ctx2, span := tracer.StartSpan(ctx1, ...)
    doWork(ctx2)
    span.Finish()
    doMoreWork(ctx1) // using prior ctx1
}
```
You may argue that someone might want to call `doMoreWork(ctx1)` from inside of `doWork(ctx2)` so that the previous ctx1 is not lexically accessible, but that would be completely non-idiomatic Go. I am not aware of a single context-related API in stdlib that allows you to "go back" to a previous context, you only go forward (usually by wrapping the previous context). If someone needs the previous context, it's on them to keep a reference to it, or structure the code so that you just return to the previous lexical scope / stack frame where previous context is active. The API `context = Tracer::EndSpan(context)` essentially breaks that by trying to go backwards via going forward (which is technically possible by storing ctx1 reference inside ctx2, but confusing as hell to me).

Reply:
I'm also wondering about the implicit scenario, where you don't care about `Context` much most of the time, as it's handled under the scenes. In Java, for example, you return nothing on `Span.end()` or `Scope.close()`. Would you **really** want to return a `Context` here?

Also, as Yuri said, you could keep a reference to the `Context` object *before* and *after* activation.

Reply:
@yurishkuro to clarify, this is not about re-activating a prior active span. It is just a clean way to write immutable code. If `tracer.EndSpan()` does not return a new context, there is no way to mark that the span is finished, except with mutation and side effects. So, when simple pseudocode with an immutable context, EndSpan must return a context with an ended span (or no span at all).

```
func (ctx1 Context) {
    ctx2 := tracer.StartSpan(ctx1, ...)
    doWork(ctx2)
    ctx2 = tracer.EndSpan(cxt2, ...)
    doMoreWork(ctx2)
    doMoreStillWork(ctx1)    // using prior ctx1
}
```

If a language makes the case that mutations and side effects are always expected, then perhaps in some places, a new context does not need to be returned. But writing it this way breaks what is otherwise a clean pattern - and implies that the "immutable" context is actually secretly mutable in some cases.

Reply:
(just to be clear, I can be swayed on whether EndSpan returns a new context. I just wanted to explain that I am not implying a prior active span should automatically become active again when a span ends.)

Reply:
I support EndSpan returning a context, since tracing is decoupled from propagation. 

Where automated context propagation is used, this may not take effect. But in manual propagation, it shouldn't be required to stop using a context simply because the span ends. The context would retain the baggage and correlations that were in effect prior to ending the span. Why might we want this? It should not be possible to call EndSpan again on that context, for example. IOW, EndSpan returns a context in which you cannot end the span again.

Reply:
> do we really have this in the spec, that End returns another context?

I think it helps to have a barebones, almost math, form of psuedo code for describing this. It can really help make clear where certain things are expected to happen.

It might be actually clearer/mathier if it were:

```
context` = Tracer::EndSpan(context)
```

This makes clear that `EndSpan` results in the span which is currently active in the `context` no longer being the active span, which means there must be a new context.

But it is certainly important in the spec to make clear that to those implementing the APIs that this isn't defining that their `EndSpan` call must actually return a context that the end user then sets to variable to continue using. I doubt it'll be an issue, but may be worth making the callout about using the common context (ThreadLocal, contextvar, etc) of each language earlier in the spec, I remember it is in there somewhere, to make sure this isn't missed.

This is exactly what the code would look like in Erlang if we didn't use the process dictionary (sort of our ThreadLocal) to store context:

```erlang
Ctx1 = ot_tracer:end_span(Ctx)
```

Since we do use the process dictionary it does not return a context to the user, but it does create a new context in the sense that it is all immutable underneath but is being tracked without the user having to do so manually, so it ends up being:

```erlang
ok = ot_tracer:end_span()
```

`ok` being an atom, not a variable, so if `end_span()` returns anything but the atom `ok` this code would cause a crash.

The barebones fake language used here may define it as `EndSpan(Context) -> Context` but the actual code, like in the case of Erlang` has a spec of `EndSpan() -> ok` because Context is implied.

Reply:
Tho maybe it should return the finished (immutable of course) span data as well? Not sure if that would be useful for anything... but like:

```
{Context`, FinishedSpan} = Tracer::EndSpan(Context)
```

Maybe useful to allow checking what the finished span was *shrug*

Reply:
> This makes clear that EndSpan results in the span which is currently active in the context no longer being the active span, which means there must be a new context.

Why must ending a span be coupled with what is active/stored in the context? I am seeing a lot of theoretical arguments, meanwhile most prior art points to not needing a context-prime.

Reply:
I don't know that it must, but I think it makes sense and simplifies things to couple them. It is certainly what a user expects when ending a span, right? When having the tracer end a span I'm telling the tracer I'm done with it.

I'd think if there were to be functions for ending a span that wasn't coupled with the active context then it wouldn't be in `Tracer` but some separate `Span` functions just for manipulating spans, separate from the active trace that the tracer is tracking.

Reply:
I don't agree that it simplifies them, I think it introduces an explicitly bad pattern. As I already explained earlier, if you want to remove the span from current context, you **go back to the previous context**, not create a new context. We should not be promoting a coding pattern that is clearly inefficient (needs a malloc), non-idiomatic (in Go), and semantically ambiguous (what becomes "active span" after that call?)

Having said that, @tedsuo - this line of pseudo-code is not even part of the proposed API functions in this OTEP.

Reply:
Ah, I'm not positive what @tedsuo had in mind, but I was not meaning there was any need for a malloc or non-idiomatic Go. Simply the span is no longer the active span in the context, this does not mean a completely new context has to be allocated and then the active span in it set to the parent.

Reply:
Yeah, each language will have to decide how to best incorporate this new API. Most will choose something other than functions operating on immutable data. 

I still do not see how it is non-idiomatic in go to return a context from `end`, but I get how it is not necessary if you assume mutable data.

So, just to clarify: I'm not proposing that the context returned from `end_span` would have a another active span in it, as if it contained a stack of spans. It would have an ended span, or no span at all. The same is true in Go today - the context would contain an ended span, correct? Only in Go, you "cheat" and mutate the contents of the supposedly immutable context.

And yes, we are just bikeshedding about examples written in pseudocode - but I do want to make sure the examples help, not hurt.